File: backend/app/services/stats_service.py
from datetime import datetime, timezone
from collections import defaultdict

from sqlalchemy.orm import Session

from app.models.entry import Entry
from app.config import TIMEZONE
import zoneinfo


def _duration_seconds(started_at: datetime, ended_at: datetime | None) -> int:
    end = ended_at if ended_at else datetime.now(timezone.utc)
    return int((end - started_at).total_seconds())


def get_stats(db: Session, user_id: int, from_dt: datetime, to_dt: datetime) -> dict:
    entries = db.query(Entry).filter(
        Entry.user_id == user_id,
        Entry.started_at >= from_dt,
        Entry.started_at <= to_dt,
    ).order_by(Entry.started_at).all()

    tz = zoneinfo.ZoneInfo(TIMEZONE)

    # Totals by label
    totals: dict[str, int] = {}
    for e in entries:
        secs = _duration_seconds(e.started_at, e.ended_at)
        totals[e.label] = totals.get(e.label, 0) + secs

    totals_by_label = [{"label": k, "total_seconds": v} for k, v in sorted(totals.items(), key=lambda x: -x[1])]

    total_all = sum(t["total_seconds"] for t in totals_by_label) or 1

    distribution = [
        {"label": t["label"], "percentage": round(t["total_seconds"] / total_all * 100, 1)}
        for t in totals_by_label
    ]

    # Tasks per day and daily series
    from collections import defaultdict
    day_counts: dict[str, int] = defaultdict(int)
    day_seconds: dict[str, int] = defaultdict(int)

    for e in entries:
        day_key = e.started_at.astimezone(tz).strftime("%Y-%m-%d")
        day_counts[day_key] += 1
        end = e.ended_at if e.ended_at else datetime.now(timezone.utc)
        day_seconds[day_key] += int((end - e.started_at).total_seconds())

    tasks_per_day = [{"date": d, "count": c} for d, c in sorted(day_counts.items())]
    daily_series = [{"date": d, "total_seconds": s} for d, s in sorted(day_seconds.items())]

    return {
        "totals_by_label": totals_by_label,
        "tasks_per_day": tasks_per_day,
        "distribution": distribution,
        "daily_series": daily_series,
    }

File: backend/src/services/stats-service.ts
import { getDb } from "../db.js";
import type { EntryRow } from "./entry-service.js";

function durationSeconds(startedAt: string, endedAt: string | null): number {
  const end = endedAt ? new Date(endedAt) : new Date();
  return Math.floor((end.getTime() - new Date(startedAt).getTime()) / 1000);
}

export function getStats(userId: number, fromDt: string, toDt: string) {
  const db = getDb();
  const stmt = db.prepare("SELECT * FROM entries WHERE user_id = ? AND started_at >= ? AND started_at <= ? ORDER BY started_at");
  stmt.bind([userId, fromDt, toDt]);
  const entries: EntryRow[] = [];
  while (stmt.step()) {
    entries.push(stmt.getAsObject() as unknown as EntryRow);
  }
  stmt.free();

  // Totals by label
  const totals: Record<string, number> = {};
  for (const e of entries) {
    const secs = durationSeconds(e.started_at, e.ended_at);
    totals[e.label] = (totals[e.label] || 0) + secs;
  }

  const totalsByLabel = Object.entries(totals)
    .map(([label, total_seconds]) => ({ label, total_seconds }))
    .sort((a, b) => b.total_seconds - a.total_seconds);

  const totalAll = totalsByLabel.reduce((s, t) => s + t.total_seconds, 0) || 1;

  const distribution = totalsByLabel.map((t) => ({
    label: t.label,
    percentage: Math.round((t.total_seconds / totalAll) * 1000) / 10,
  }));

  // Tasks per day and daily series
  const dayCounts: Record<string, number> = {};
  const daySeconds: Record<string, number> = {};

  for (const e of entries) {
    const dayKey = e.started_at.slice(0, 10);
    dayCounts[dayKey] = (dayCounts[dayKey] || 0) + 1;
    const end = e.ended_at ? new Date(e.ended_at) : new Date();
    daySeconds[dayKey] = (daySeconds[dayKey] || 0) + Math.floor((end.getTime() - new Date(e.started_at).getTime()) / 1000);
  }

  const tasksPerDay = Object.entries(dayCounts)
    .map(([date, count]) => ({ date, count }))
    .sort((a, b) => a.date.localeCompare(b.date));

  const dailySeries = Object.entries(daySeconds)
    .map(([date, total_seconds]) => ({ date, total_seconds }))
    .sort((a, b) => a.date.localeCompare(b.date));

  return {
    totals_by_label: totalsByLabel,
    tasks_per_day: tasksPerDay,
    distribution,
    daily_series: dailySeries,
  };
}